File: demo/ffi/ygui2/typescript/catalog.ts
// ygui2 from TypeScript — the widget catalog (a ygreeter2 port).
// RUNNABLE inside a yetty pane:
//
//     cd demo/ffi/ygui2/typescript && npm install
//     node catalog.ts
//
// Every phase-6 widget wired from TypeScript: chips, a toggle driving an
// overlay tooltip, a radio group driving a stepper, slider→progress
// binding, a spinner, a textinput that greets on Enter, a dropdown, a
// dialog in the overlay, and a statusbar mirroring every event.
// Tab/Shift-Tab walk focus, Esc closes overlays, Ctrl-C quits (also `q`
// while no text input holds focus).
import { App, RadioGroup, type Node } from "@yetty/ydraw/ygui2";

const app = new App();

const column = app.root.column({ grow: 1, gap: 10, pad: 16 });

const titleRow = column.row({ basis: 28, gap: 10 });
titleRow.label({ text: "ygui2 catalog — TypeScript edition", fg: "#74C5A5", basis: 260 });
["drawable", "contract", "toolkit"].forEach((chipText, index) => {
  titleRow.chip({ label: chipText, selectable: true, selected: index === 0,
    basis: 76, cross: 22 });
});

column.separator({ basis: 8 });

let status: Node | null = null; // created last; the closures capture the slot

function show(text: string): void {
  if (status !== null) {
    status.status({ left: text });
  }
}

const tooltip = app.tooltip({ text: "the toggle controls me", x: 150, y: 66 });

const switchRow = column.row({ basis: 28, gap: 10 });
switchRow.label({ text: "switches", fg: "#9FA7A8", basis: 110 });
switchRow.toggle({ label: "tooltip", basis: 120,
  onToggle: (node: Node) => {
    const checked = node.toggleChecked();
    tooltip.setVisible(checked);
    show(checked ? "toggle: on" : "toggle: off");
  } });
let stepper: Node | null = null;
const group = new RadioGroup();
for (let option = 0; option < 3; option++) {
  switchRow.radio({ label: `opt ${option + 1}`, group, selected: option === 0,
    basis: 90,
    onSelect: (index: number) => {
      stepper?.stepperCurrent(index);
      show(`radio: option ${index + 1}`);
    } });
}
stepper = switchRow.stepper({ count: 3, current: 0, basis: 80 });

const valueRow = column.row({ basis: 28, gap: 10 });
valueRow.label({ text: "values", fg: "#9FA7A8", basis: 110 });
let bar: Node | null = null;
valueRow.slider({ value: 0.35, basis: 160,
  onChange: (node: Node) => {
    const value = node.sliderValue();
    bar?.setValue(value);
    show(`slider: ${Math.round(value * 100)}%`);
  } });
bar = valueRow.progress({ value: 0.35, basis: 140, cross: 12 });
valueRow.spinner({ value: 3, minimum: 0, maximum: 10, step: 1, basis: 110,
  onChange: (node: Node) => show(`spinner: ${node.spinnerValue()}`) });

const entryRow = column.row({ basis: 28, gap: 10 });
entryRow.label({ text: "entry", fg: "#9FA7A8", basis: 110 });
entryRow.textinput({ placeholder: "type a name, Enter greets", basis: 180,
  onSubmit: (node: Node) => show(`hello, ${node.inputText() || "stranger"}`) });

entryRow.dropdown({ items: ["plasma", "aurora", "nebula"], basis: 130,
  onChange: (index: number) => show(`dropdown: item ${index + 1}`) });

const dialog = app.dialog({ title: "about the catalog", x: 140, y: 90, width: 300, height: 150,
  onClose: () => show("dialog closed") });
dialog.label({ text: "every widget, one wire contract", basis: 20 });
entryRow.button({ label: "open dialog", basis: 110,
  onClick: () => {
    dialog.setVisible(true);
    show("dialog opened");
  } });

column.column({ grow: 1.0 });
status = column.statusbar({ left: "ready",
  right: "Tab: focus  Esc: close  Ctrl-C: quit", basis: 24 });

await app.run();

File: demo/ffi/ygui2/lua/catalog.lua
-- ygui2 from Lua — the widget catalog (a ygreeter2 port). RUNNABLE inside
-- a yetty pane:
--
--     LUA_PATH="bindings/lua/?.lua;;" luajit demo/ffi/ygui2/lua/catalog.lua
--
-- Every phase-6 widget wired from Lua: chips, a toggle driving an overlay
-- tooltip, a radio group driving a stepper, slider→progress binding, a
-- spinner, a textinput that greets on Enter, a dropdown, a dialog in the
-- overlay, and a statusbar mirroring every event. Tab/Shift-Tab walk
-- focus, Esc closes overlays, Ctrl-C quits (also `q` while no text input
-- holds focus).
local ygui2 = require("yetty.ygui2")

local app = ygui2.App()

local column = app.root:column{grow = 1, gap = 10, pad = 16}

local title_row = column:row{basis = 28, gap = 10}
title_row:label{text = "ygui2 catalog — Lua edition", fg = "#74C5A5", basis = 260}
for index, chip_text in ipairs({"drawable", "contract", "toolkit"}) do
  title_row:chip{label = chip_text, selectable = true, selected = index == 1,
                 basis = 76, cross = 22}
end

column:separator{basis = 8}

local status = nil -- created last; the closures capture the slot

local function show(text)
  if status ~= nil then
    status:status{left = text}
  end
end

local tooltip = app:tooltip{text = "the toggle controls me", x = 150, y = 66}

local switch_row = column:row{basis = 28, gap = 10}
switch_row:label{text = "switches", fg = "#9FA7A8", basis = 110}
switch_row:toggle{label = "tooltip", basis = 120,
                  on_toggle = function(node)
                    local checked = node:toggle_checked()
                    tooltip:set_visible(checked)
                    show(checked and "toggle: on" or "toggle: off")
                  end}
local stepper = nil
local group = ygui2.RadioGroup()
for option = 0, 2 do
  switch_row:radio{label = "opt " .. (option + 1), group = group, selected = option == 0,
                   basis = 90,
                   on_select = function(index)
                     stepper:stepper_current(index)
                     show("radio: option " .. (index + 1))
                   end}
end
stepper = switch_row:stepper{count = 3, current = 0, basis = 80}

local value_row = column:row{basis = 28, gap = 10}
value_row:label{text = "values", fg = "#9FA7A8", basis = 110}
local bar = nil
value_row:slider{value = 0.35, basis = 160,
                 on_change = function(node)
                   local value = node:slider_value()
                   bar:set_value(value)
                   show(string.format("slider: %.0f%%", value * 100))
                 end}
bar = value_row:progress{value = 0.35, basis = 140, cross = 12}
value_row:spinner{value = 3, minimum = 0, maximum = 10, step = 1, basis = 110,
                  on_change = function(node)
                    show(string.format("spinner: %g", node:spinner_value()))
                  end}

local entry_row = column:row{basis = 28, gap = 10}
entry_row:label{text = "entry", fg = "#9FA7A8", basis = 110}
entry_row:textinput{placeholder = "type a name, Enter greets", basis = 180,
                    on_submit = function(node)
                      local name = node:input_text()
                      show("hello, " .. (#name > 0 and name or "stranger"))
                    end}

entry_row:dropdown{items = {"plasma", "aurora", "nebula"}, basis = 130,
                   on_change = function(index) show("dropdown: item " .. (index + 1)) end}

local dialog = app:dialog{title = "about the catalog", x = 140, y = 90, width = 300, height = 150,
                          on_close = function() show("dialog closed") end}
dialog:label{text = "every widget, one wire contract", basis = 20}
entry_row:button{label = "open dialog", basis = 110,
                 on_click = function()
                   dialog:set_visible(true)
                   show("dialog opened")
                 end}

column:column{grow = 1.0}
status = column:statusbar{left = "ready",
                          right = "Tab: focus  Esc: close  Ctrl-C: quit", basis = 24}

app:run()
